src/include/sof/audio/audio_stream.h

```diff
@@ -220,6 +220,38 @@ static inline void audio_stream_init(struct audio_stream *buffer,
 	audio_stream_reset(buffer);
 }
 
+static inline void audio_stream_invalidate(struct audio_stream *buffer,
+					   uint32_t bytes)
+{
+	uint32_t head = bytes;
+	uint32_t tail = 0;
+
+	if ((char *)buffer->r_ptr + bytes > (char *)buffer->end_addr) {
+		head = (char *)buffer->end_addr - (char *)buffer->r_ptr;
+		tail = bytes - head;
+	}
+
+	dcache_invalidate_region(buffer->r_ptr, head);
+	if (tail)
+		dcache_invalidate_region(buffer->addr, tail);
+}
+
+static inline void audio_stream_writeback(struct audio_stream *buffer,
+					  uint32_t bytes)
+{
+	uint32_t head = bytes;
+	uint32_t tail = 0;
+
+	if ((char *)buffer->w_ptr + bytes > (char *)buffer->end_addr) {
+		head = (char *)buffer->end_addr - (char *)buffer->w_ptr;
+		tail = bytes - head;
+	}
+
+	dcache_writeback_region(buffer->w_ptr, head);
+	if (tail)
+		dcache_writeback_region(buffer->addr, tail);
+}
```

Comment:
Btw, I would rename head/tail to include "size" as I initially read this code as some list iteration...

Reply:
Done.